knora/dsplib/utils/stack_handling.py

```diff
@@ -0,0 +1,122 @@
+import json
+import re
+import subprocess
+import time
+from pathlib import Path
+from typing import Optional
+
+import requests
+
+from knora.dsplib.models.helpers import BaseError
+
+# relative path to "knora/dsplib/docker", to make it accessible when dsp-tools is called from another working directory
+docker_path = Path(__file__).parent / Path("../docker")
+
+
+def start_stack(
```

Comment:
a remark purely on code style: a function should never be 100 lines long (including docstring), it should never "do more than one thing" and it should normally not have to operate on different levels of abstraction. All three of those are the case for this function.  
If you find yourself writing that kind of code (which is normal, that's how one writes code), then you should afterwards refactor it by extracting all sub-tasks and lower-level stuff into separate private functions.  
(I'm freely quoting Bob Martin's "Clean Architecture" here - but it's a good rule of thumb to follow.)

Reply:
Thanks for the hint, it's important that I learn such stuff about architecture. I think I won't do it in this PR. But a general question concerning this rule of thumb: Doesn't it make the code more complicated if I have a lot of private methods that are so specific that are only called once?

Reply:
Bob Martin (and I) would say: no, your code should get more readable and easier to understand.

You have to get some things right though:  
- a method name should be descriptive of what the method does
- a method must only operate on one abstraction level, namely calling methods of the next lower abstraction level

(and there are more finesses of course)

If you follow those rules, all your methods should turn out pretty much self explanatory: the name says what it does and the (maybe 3 to 10) function calls within it are self-explaining steps that need to be taken, to achieve this.  
And if you are interested in further detail, you can jump into each of those steps, where the same thing will be true: the function name is what it does, and the contents are the steps taken to do it.

Does that make sense?  
If you're interested, there is - among tons of other material - a several hour long talk series by "Uncle Bob" named "Clean Code" on Youtube, which is definitely worth watching.